File: app/src/main/java/com/example/purepawapp/ui/cart/CartAdapter.java
package com.example.purepawapp.ui.cart;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.example.purepawapp.data.model.CartItem;
import com.example.purepawapp.databinding.ItemCartBinding;
import com.example.purepawapp.util.CurrencyUtils;

import java.util.function.Consumer;

public class CartAdapter extends ListAdapter<CartItem, CartAdapter.CartViewHolder> {

    private final Consumer<CartItem> onIncrease;
    private final Consumer<CartItem> onDecrease;
    private final Consumer<CartItem> onRemove;

    private static final DiffUtil.ItemCallback<CartItem> DIFF_CALLBACK = new DiffUtil.ItemCallback<>() {
        @Override
        public boolean areItemsTheSame(@NonNull CartItem oldItem, @NonNull CartItem newItem) {
            return oldItem.getProduct().getId().equals(newItem.getProduct().getId())
                    && oldItem.getVariant().getId().equals(newItem.getVariant().getId());
        }

        @Override
        public boolean areContentsTheSame(@NonNull CartItem oldItem, @NonNull CartItem newItem) {
            return oldItem.getQuantity() == newItem.getQuantity();
        }
    };

    public CartAdapter(Consumer<CartItem> onIncrease, Consumer<CartItem> onDecrease, Consumer<CartItem> onRemove) {
        super(DIFF_CALLBACK);
        this.onIncrease = onIncrease;
        this.onDecrease = onDecrease;
        this.onRemove = onRemove;
    }

    @NonNull
    @Override
    public CartViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemCartBinding binding = ItemCartBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new CartViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull CartViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    public class CartViewHolder extends RecyclerView.ViewHolder {
        private final ItemCartBinding binding;

        CartViewHolder(ItemCartBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(CartItem item) {
            binding.tvName.setText(item.getProduct().getName());
            binding.tvVariant.setText(item.getVariant().getName());
            binding.tvPrice.setText(CurrencyUtils.toVndString(item.getUnitPrice()));
            binding.tvQuantity.setText(String.valueOf(item.getQuantity()));
            binding.btnMinus.setOnClickListener(v -> onDecrease.accept(item));
            binding.btnPlus.setOnClickListener(v -> onIncrease.accept(item));
            binding.btnRemove.setOnClickListener(v -> onRemove.accept(item));
        }
    }
}

File: app/src/main/java/com/example/purepawapp/ui/cart/CartAdapter.kt
package com.example.purepawapp.ui.cart

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.purepawapp.data.model.CartItem
import com.example.purepawapp.databinding.ItemCartBinding
import com.example.purepawapp.util.toVndString

class CartAdapter(
    private val onIncrease: (CartItem) -> Unit,
    private val onDecrease: (CartItem) -> Unit,
    private val onRemove: (CartItem) -> Unit
) : ListAdapter<CartItem, CartAdapter.CartViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CartViewHolder {
        val binding = ItemCartBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return CartViewHolder(binding)
    }

    override fun onBindViewHolder(holder: CartViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class CartViewHolder(private val binding: ItemCartBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(item: CartItem) {
            binding.tvName.text = item.product.name
            binding.tvVariant.text = item.variant.name
            binding.tvPrice.text = item.unitPrice.toVndString()
            binding.tvQuantity.text = item.quantity.toString()
            binding.btnMinus.setOnClickListener { onDecrease(item) }
            binding.btnPlus.setOnClickListener { onIncrease(item) }
            binding.btnRemove.setOnClickListener { onRemove(item) }
        }
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<CartItem>() {
            override fun areItemsTheSame(oldItem: CartItem, newItem: CartItem): Boolean =
                oldItem.product.id == newItem.product.id && oldItem.variant.id == newItem.variant.id

            override fun areContentsTheSame(oldItem: CartItem, newItem: CartItem): Boolean =
                oldItem == newItem
        }
    }
}
